quartz/util/encryption.ts

```diff
@@ -0,0 +1,527 @@
+  password: string,
+  salt: ArrayBuffer | string | undefined = undefined,
+): Promise<Hash> {
+  const passwordBytes = stringToArrayBuffer(password)
+
+  let saltBytes: Uint8Array | null = null
+
+  if (typeof salt === "string") {
+    saltBytes = new Uint8Array(hexToArrayBuffer(salt))
+  } else if (salt !== undefined) {
+    saltBytes = new Uint8Array(salt)
+  } else {
+    saltBytes = crypto.getRandomValues(new Uint8Array(16))
+  }
+
+  const combined = new Uint8Array(passwordBytes.byteLength + saltBytes.byteLength)
+  combined.set(new Uint8Array(passwordBytes), 0)
+  combined.set(saltBytes, passwordBytes.byteLength)
+
+  const hashBuffer = await crypto.subtle.digest("SHA-256", combined)
+  return {
+    hash: arrayBufferToHex(hashBuffer),
+    salt: arrayBufferToHex(saltBytes.buffer as ArrayBuffer),
+  }
+}
+
+export async function verifyPasswordHash(
+  password: string,
+  passwordHashData: Hash,
+): Promise<boolean> {
+  const { hash: passwordHash } = await hashString(password, passwordHashData.salt)
+  return passwordHash === passwordHashData.hash
+}
+
+export async function encryptContent(
+  content: string,
+  password: string,
+  config: CompleteCryptoConfig,
+): Promise<EncryptionResult> {
+  checkCryptoSupport()
+
+  const { algorithm } = config
+
+  if (!SUPPORTED_ALGORITHMS.includes(algorithm as SupportedEncryptionAlgorithm)) {
+    throw new Error(
+      `Unsupported encryption algorithm: ${algorithm}. Supported: ${SUPPORTED_ALGORITHMS.join(", ")}`,
+    )
+  }
+
+  // Generate random salt for encryption
+  const initializationVector = crypto.getRandomValues(new Uint8Array(16))
+
+  // Create encryption hash and derive key
+  const encryptionHashData = await hashString(password)
+  const key = await deriveKeyFromHash(encryptionHashData.hash, algorithm)
+
+  // Prepare content for encryption
+  const contentBuffer = stringToArrayBuffer(content)
+
+  let encryptedBuffer: ArrayBuffer
+  let authTag: ArrayBuffer | undefined
+
+  try {
+    if (algorithm === "aes-256-gcm") {
+      // GCM mode - the Web Crypto API returns ciphertext with auth tag appended
+      const encryptedWithTag = await crypto.subtle.encrypt(
+        {
+          name: "AES-GCM",
+          iv: initializationVector,
+        },
+        key,
+        contentBuffer,
+      )
+
+      // The last 16 bytes (128 bits) are the authentication tag
+      const encryptedBytes = new Uint8Array(encryptedWithTag)
+      const ciphertext = encryptedBytes.slice(0, -16)
+      const authTagBytes = encryptedBytes.slice(-16)
+
+      encryptedBuffer = ciphertext.buffer
+      authTag = authTagBytes.buffer
+    } else if (algorithm === "aes-256-cbc") {
+      // CBC mode
+      encryptedBuffer = await crypto.subtle.encrypt(
+        {
+          name: "AES-CBC",
+          iv: initializationVector,
+        },
+        key,
+        contentBuffer,
+      )
+    } else {
+      throw new Error("Unsupported algorithm: " + algorithm)
+    }
+  } catch (error) {
```

Comment:
this seems too complicated imo. We can just use one type of encryption with salt hash should be more than good enough.